### DISCUSSION TITLE
What does OOK_PWM modulation mean?  OOK (ASK) refers to amplitude and PWM refers to width?

### DISCUSSION THREAD

**Discussion by sofakng** [merbanan/rtl_433#3101] [General]
I'm trying to understand how RTL_433 is decoding (demodulating) signals.  The signal I'm decoding is definitely OOK (ASK) but RTL_433 seems to refer to this as OOK_PWM.

What does this mean?

I thought OOK (ASK) refers to amplitude ("height") and PWM refers to pulse width ("width") so I don't see how they are related?

**Comment by merbanan** [comment]
You can have several ways to represent the message bits. One way is to map it to long and short pulses (PWM) and another is to map it to long and short gaps between pulses (PPM). Basically rtl_433 rf signal decoding translates the signal to a long list of pulses in time and then it tries to decode all the protocols it knows against this pulse list. If something matches a protocol it outputs a decoded message.

**Reply by zuckschwerdt** [reply]
If you missed the answer: there is an introduction to pulse formats with rtl_433 terms here: https://triq.org/rtl_433/PULSE_FORMATS.html

(Binary-)PWM is just to denote that there is no fixed bit-clock (PCM) but the actual length of a pulse (or gap: PPM) encodes the bits.

**Reply by gcormier** [reply]
I figured I'd jump on this discussion since it relates to my head scratching.

Regarding #3034 

Is this just a regular OOK PWM signal?

![image](https://github.com/user-attachments/assets/819a494a-73c7-424e-aa7d-8085538a6f49)

I'm trying to code the respective transmitter now and the different duration pulses are confusing me. I was expecting to see either

1. The same short pulse, with two different HIGH pulse widths for 1/0 or,
2. The same long pulse, with two different LOW pulse widths for 1/0

**Comment by zuckschwerdt** [comment]
It looks like neither PWM with leading or trailing gap and also not like MC. We once had something like this an dubbed it PIWM.
You'd need multiple codes as PCM to find a pattern. PCM for this e.g. `111010101010111010010011011011011011` (or in PDV: https://triq.org/pdv/#AAB104050D01A40384068F8191919191819292A1A1A1A1A355)

**Reply by gcormier** [reply]
Ok phew, makes me feel a bit better. I think in this case for a transmitter, it would be best just to brute-force (replay) the transmission.

**Reply by zuckschwerdt** [reply]
PCM ~433 µs (2.3 kbps) should work then.

**Reply by gcormier** [reply]
On second thought, is this fixed bit width PWM?

![image](https://github.com/user-attachments/assets/bbb1f6ae-5d0e-461e-afb6-25a0a6e496e3)

**Reply by zuckschwerdt** [reply]
Only if you disregard the first pulses as warmup/preamble and then take both three units long pulses as sync, i.e. "Fixed ID" is not part of the code. Is every pulse seen just one bit, i.e. is a message only 7 bit?

**Reply by gcormier** [reply]
Yup it's just those 7 bits.

I fixed my transmission for the first 6 bits, and then I made the last 7 fixed width, and that seems to have done the trick!
